File: tests/filter.test.js
import { filterTorrents } from '../src/filter.js'

const t = (overrides = {}) => ({
  title: overrides.title ?? 'Some Movie 2024 1080p WEB-DL x264',
  parsedTitle: {
    quality: 'quality' in overrides ? overrides.quality : '1080p',
    source: overrides.source ?? 'WEB-DL',
    hdr: overrides.hdr ?? null,
    audio: overrides.audio ?? null,
  },
  size: 'size' in overrides ? overrides.size : 2 * 1024 ** 3, // 2 GB default
  seeders: overrides.seeders ?? 10,
})

const GB = 1024 ** 3
const MB = 1024 ** 2

describe('filterTorrents', () => {
  test('no config options → everything passes through', () => {
    const items = [
      t({ quality: '4K' }),
      t({ quality: '1080p' }),
      t({ quality: '720p' }),
      t({ quality: '480p' }),
      t({ quality: 'CAM' }),
      t({ quality: null }),
    ]
    expect(filterTorrents(items, {})).toHaveLength(items.length)
  })

  test("maxQuality: '720p' drops 1080p and 4K, keeps 720p, 480p, null", () => {
    const items = [
      t({ quality: '4K' }),
      t({ quality: '1080p' }),
      t({ quality: '720p' }),
      t({ quality: '480p' }),
      t({ quality: null }),
    ]
    const out = filterTorrents(items, { maxQuality: '720p' })
    const qualities = out.map(x => x.parsedTitle.quality)
    expect(qualities).toEqual(['720p', '480p', null])
  })

  test("minQuality: '720p' drops 480p and null, keeps 720p+", () => {
    const items = [
      t({ quality: '4K' }),
      t({ quality: '1080p' }),
      t({ quality: '720p' }),
      t({ quality: '480p' }),
      t({ quality: null }),
    ]
    const out = filterTorrents(items, { minQuality: '720p' })
    const qualities = out.map(x => x.parsedTitle.quality)
    expect(qualities).toEqual(['4K', '1080p', '720p'])
  })

  test('maxSizeGb: 10 drops >10 GB', () => {
    const items = [
      t({ size: 5 * GB }),
      t({ size: 10 * GB }), // on the boundary — kept (not strictly greater)
      t({ size: 15 * GB }),
    ]
    const out = filterTorrents(items, { maxSizeGb: 10 })
    expect(out.map(x => x.size)).toEqual([5 * GB, 10 * GB])
  })

  test('minSizeMb: 500 drops <500 MB', () => {
    const items = [
      t({ size: 100 * MB }),
      t({ size: 500 * MB }), // on the boundary — kept (not strictly less)
      t({ size: 900 * MB }),
    ]
    const out = filterTorrents(items, { minSizeMb: 500 })
    expect(out.map(x => x.size)).toEqual([500 * MB, 900 * MB])
  })

  test("blacklistKeywords drops matching titles, word-boundary (SCAM not matched by CAM)", () => {
    const items = [
      t({ title: 'Movie CAM Rip' }),
      t({ title: 'Movie HDCAM 2024' }),
      t({ title: 'A SCAM Story' }),
      t({ title: 'Clean Title 1080p' }),
    ]
    const out = filterTorrents(items, { blacklistKeywords: ['CAM', 'HDCAM'] })
    expect(out.map(x => x.title)).toEqual(['A SCAM Story', 'Clean Title 1080p'])
  })

  test("regex injection: blacklist like ['.*'] is escaped and doesn't match everything", () => {
    const items = [
      t({ title: 'Movie 1080p' }),
      t({ title: 'Literal .* match' }),
    ]
    const out = filterTorrents(items, { blacklistKeywords: ['.*'] })
    // The first title does NOT contain ".*" literally, so must not be dropped.
    expect(out.some(x => x.title === 'Movie 1080p')).toBe(true)
  })

  test('unknown size passes through size filters (size not finite)', () => {
    const items = [t({ size: undefined }), t({ size: null })]
    const out = filterTorrents(items, { maxSizeGb: 1, minSizeMb: 50 })
    expect(out).toHaveLength(2)
  })

  test('combined filters compose', () => {
    const items = [
      t({ quality: '4K', size: 30 * GB, title: 'Big 4K' }),
      t({ quality: '1080p', size: 5 * GB, title: 'Good 1080p' }),
      t({ quality: '720p', size: 50 * MB, title: 'Tiny 720p' }),
      t({ quality: '480p', size: 1 * GB, title: 'Old 480p' }),
    ]
    const out = filterTorrents(items, {
      maxQuality: '1080p',
      minQuality: '720p',
      maxSizeGb: 10,
      minSizeMb: 100,
    })
    expect(out.map(x => x.title)).toEqual(['Good 1080p'])
  })
})

File: internal/filter/filter_test.go
package filter

import (
	"testing"

	"github.com/lborruto/jackstream/internal/config"
	"github.com/lborruto/jackstream/internal/parser"
)

type tItem struct {
	T       string
	Parsed  parser.Parsed
	SizeVal int64
}

func (t tItem) Title() string              { return t.T }
func (t tItem) ParsedTitle() parser.Parsed { return t.Parsed }
func (t tItem) Size() int64                { return t.SizeVal }

func mk(title, quality string, sizeBytes int64) tItem {
	return tItem{T: title, Parsed: parser.Parsed{Quality: quality}, SizeVal: sizeBytes}
}

func titles(fs []Filterable) []string {
	out := make([]string, len(fs))
	for i, f := range fs {
		out[i] = f.Title()
	}
	return out
}

func eq(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestEmptyConfigPassesEverything(t *testing.T) {
	in := []Filterable{mk("a", "1080p", 1<<30), mk("b", "720p", 500<<20), mk("c", "CAM", 1<<29)}
	got := Filter(in, config.Config{})
	if len(got) != 3 {
		t.Errorf("want 3, got %d", len(got))
	}
}

func TestMaxQualityDropsHigher(t *testing.T) {
	in := []Filterable{
		mk("4k", "4K", 1<<30),
		mk("hd", "1080p", 1<<30),
		mk("sd", "720p", 1<<30),
		mk("unk", "", 1<<30),
	}
	got := Filter(in, config.Config{MaxQuality: "720p"})
	want := []string{"sd", "unk"}
	if !eq(titles(got), want) {
		t.Errorf("max=720p want %v, got %v", want, titles(got))
	}
}

func TestMinQualityDropsLower(t *testing.T) {
	in := []Filterable{
		mk("4k", "4K", 1<<30),
		mk("hd", "1080p", 1<<30),
		mk("sd", "720p", 1<<30),
		mk("vhs", "480p", 1<<30),
		mk("unk", "", 1<<30),
	}
	got := Filter(in, config.Config{MinQuality: "720p"})
	want := []string{"4k", "hd", "sd"}
	if !eq(titles(got), want) {
		t.Errorf("min=720p want %v, got %v", want, titles(got))
	}
}

func TestMaxSizeGbDropsLarger(t *testing.T) {
	in := []Filterable{mk("ok", "1080p", 5<<30), mk("big", "1080p", 20<<30)}
	got := Filter(in, config.Config{MaxSizeGb: 10})
	if len(got) != 1 || got[0].Title() != "ok" {
		t.Errorf("want [ok], got %v", titles(got))
	}
}

func TestMinSizeMbDropsSmaller(t *testing.T) {
	in := []Filterable{mk("tiny", "1080p", 100<<20), mk("ok", "1080p", 2<<30)}
	got := Filter(in, config.Config{MinSizeMb: 500})
	if len(got) != 1 || got[0].Title() != "ok" {
		t.Errorf("want [ok], got %v", titles(got))
	}
}

func TestBlacklistWordBoundary(t *testing.T) {
	in := []Filterable{
		mk("Movie.2024.1080p.CAM.x264", "1080p", 1<<30),
		mk("Scam.Likely.2024.1080p.WEB-DL", "1080p", 1<<30),
	}
	got := Filter(in, config.Config{BlacklistKeywords: []string{"CAM"}})
	if len(got) != 1 || got[0].Title() != "Scam.Likely.2024.1080p.WEB-DL" {
		t.Errorf("blacklist dropped wrong item: %v", titles(got))
	}
}

func TestBlacklistMultipleKeywords(t *testing.T) {
	in := []Filterable{
		mk("Cam.2024", "CAM", 1<<30),
		mk("HDCAM.2024", "CAM", 1<<30),
		mk("Clean.2024", "1080p", 1<<30),
	}
	got := Filter(in, config.Config{BlacklistKeywords: []string{"CAM", "HDCAM"}})
	if len(got) != 1 || got[0].Title() != "Clean.2024" {
		t.Errorf("want [Clean.2024], got %v", titles(got))
	}
}

func TestBlacklistRegexInjectionEscaped(t *testing.T) {
	in := []Filterable{mk("Movie.2024.1080p", "1080p", 1<<30)}
	got := Filter(in, config.Config{BlacklistKeywords: []string{".*"}})
	if len(got) != 1 {
		t.Errorf("`.*` must be escaped, should not match everything")
	}
}

func TestBlacklistCaseInsensitive(t *testing.T) {
	in := []Filterable{mk("Movie.2024.HdCaM.x264", "", 1<<30)}
	got := Filter(in, config.Config{BlacklistKeywords: []string{"hdcam"}})
	if len(got) != 0 {
		t.Errorf("blacklist must be case-insensitive, got %v", titles(got))
	}
}
